File: implementations/fsharp/src/Skight.AgentPlatform.FSharp/ToolSecurity.fs
namespace Skight.AgentPlatform.FSharp

open System
open System.IO
open System.Text

module ToolSecurity =
    let private comparison =
        if Environment.OSVersion.Platform = PlatformID.Win32NT then
            StringComparison.OrdinalIgnoreCase
        else
            StringComparison.Ordinal

    let private normalizeRootPath (workspaceRoot: string) =
        let canonical = Path.GetFullPath(workspaceRoot).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)
        canonical + string Path.DirectorySeparatorChar

    let private canonicalizeTargetPath (workspaceRoot: string) (targetPath: string) =
        if Path.IsPathRooted(targetPath) then
            Path.GetFullPath(targetPath)
        else
            Path.GetFullPath(Path.Combine(workspaceRoot, targetPath))

    /// Validates that path is inside the workspace sandbox.
    let validatePathInSandbox (workspaceRoot: string) (targetPath: string) : Result<string, string> =
        try
            if String.IsNullOrWhiteSpace(workspaceRoot) then
                Error "Workspace root cannot be empty."
            elif String.IsNullOrWhiteSpace(targetPath) then
                Error "Target path cannot be empty."
            else
                let rootWithSeparator = normalizeRootPath workspaceRoot
                let targetCanonical = canonicalizeTargetPath workspaceRoot targetPath
                let targetWithSeparator = targetCanonical + string Path.DirectorySeparatorChar

                if targetWithSeparator.StartsWith(rootWithSeparator, comparison) then
                    Ok targetCanonical
                else
                    Error (sprintf "Access denied: Path '%s' is outside workspace sandbox." targetPath)
        with ex ->
            Error (sprintf "Invalid path '%s': %s" targetPath ex.Message)

    /// Backward-compatible alias used by design drafts.
    let validateSandboxPath (workspaceRoot: string) (targetPath: string) : Result<string, string> =
        validatePathInSandbox workspaceRoot targetPath

    let truncateOutputWithLimits (maxBytes: int) (maxLines: int) (output: string) : string =
        if String.IsNullOrEmpty(output) || maxBytes <= 0 || maxLines <= 0 then
            output
        else
            let normalized = output.Replace("\r\n", "\n")
            let splitLines = normalized.Split('\n')

            let lineTrimmed, lineTruncatedBy =
                if splitLines.Length > maxLines then
                    let kept = splitLines |> Array.take maxLines |> String.concat "\n"
                    kept, splitLines.Length - maxLines
                else
                    normalized, 0

            let withLineMarker =
                if lineTruncatedBy > 0 then
                    sprintf "%s\n\n[Output truncated: %d lines hidden to protect context window]" lineTrimmed lineTruncatedBy
                else
                    lineTrimmed

            let currentBytes = Encoding.UTF8.GetByteCount(withLineMarker)
            if currentBytes <= maxBytes then
                withLineMarker
            else
                let marker = sprintf "\n\n[Output truncated: exceeds %d bytes safety limit]" maxBytes
                let markerBytes = Encoding.UTF8.GetByteCount(marker)
                let allowedBytes = max 0 (maxBytes - markerBytes)

                let rec findMaxChars low high =
                    if low >= high then low
                    else
                        let mid = (low + high + 1) / 2
                        let bytes = Encoding.UTF8.GetByteCount(withLineMarker.Substring(0, mid))
                        if bytes <= allowedBytes then findMaxChars mid high
                        else findMaxChars low (mid - 1)

                let length = findMaxChars 0 withLineMarker.Length
                withLineMarker.Substring(0, length) + marker

    /// Truncates large tool outputs by bytes and line count.
    let truncateOutput (maxBytes: int) (output: string) : string =
        truncateOutputWithLimits maxBytes 500 output

File: implementations/csharp/src/Skight.AgentPlatform/ToolSecurity.cs
using System;
using System.IO;
using System.Text;

namespace Skight.AgentPlatform
{
    public static class ToolSecurity
    {
        private static readonly StringComparison Comparison = 
            Environment.OSVersion.Platform == PlatformID.Win32NT ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal;

        private static string NormalizeRootPath(string workspaceRoot)
        {
            var canonical = Path.GetFullPath(workspaceRoot).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            return canonical + Path.DirectorySeparatorChar;
        }

        private static string CanonicalizeTargetPath(string workspaceRoot, string targetPath)
        {
            return Path.IsPathRooted(targetPath) 
                ? Path.GetFullPath(targetPath) 
                : Path.GetFullPath(Path.Combine(workspaceRoot, targetPath));
        }

        public static string ValidatePathInSandbox(string workspaceRoot, string targetPath)
        {
            if (string.IsNullOrWhiteSpace(workspaceRoot)) throw new ArgumentException("Workspace root cannot be empty.");
            if (string.IsNullOrWhiteSpace(targetPath)) throw new ArgumentException("Target path cannot be empty.");

            var rootWithSeparator = NormalizeRootPath(workspaceRoot);
            var targetCanonical = CanonicalizeTargetPath(workspaceRoot, targetPath);
            var targetWithSeparator = targetCanonical + Path.DirectorySeparatorChar;

            if (targetWithSeparator.StartsWith(rootWithSeparator, Comparison))
            {
                return targetCanonical;
            }
            
            throw new UnauthorizedAccessException($"Access denied: Path '{targetPath}' is outside workspace sandbox.");
        }

        public static string TruncateOutputWithLimits(int maxBytes, int maxLines, string output)
        {
            if (string.IsNullOrEmpty(output) || maxBytes <= 0 || maxLines <= 0) return output;

            var normalized = output.Replace("\r\n", "\n");
            var splitLines = normalized.Split('\n');
            
            string lineTrimmed;
            int lineTruncatedBy;
            
            if (splitLines.Length > maxLines)
            {
                var kept = new string[maxLines];
                Array.Copy(splitLines, kept, maxLines);
                lineTrimmed = string.Join("\n", kept);
                lineTruncatedBy = splitLines.Length - maxLines;
            }
            else
            {
                lineTrimmed = normalized;
                lineTruncatedBy = 0;
            }

            var withLineMarker = lineTruncatedBy > 0 
                ? $"{lineTrimmed}\n\n[Output truncated: {lineTruncatedBy} lines hidden to protect context window]" 
                : lineTrimmed;

            var currentBytes = Encoding.UTF8.GetByteCount(withLineMarker);
            if (currentBytes <= maxBytes) return withLineMarker;

            var marker = $"\n\n[Output truncated: exceeds {maxBytes} bytes safety limit]";
            var markerBytes = Encoding.UTF8.GetByteCount(marker);
            var allowedBytes = Math.Max(0, maxBytes - markerBytes);
            
            int low = 0, high = withLineMarker.Length;
            while (low < high)
            {
                int mid = (low + high + 1) / 2;
                if (Encoding.UTF8.GetByteCount(withLineMarker.Substring(0, mid)) <= allowedBytes)
                    low = mid;
                else
                    high = mid - 1;
            }
            
            return withLineMarker.Substring(0, low) + marker;
        }

        public static string TruncateOutput(int maxBytes, string output) => TruncateOutputWithLimits(maxBytes, 500, output);
    }
}
